corehq/form_processor/submission_post.py

```diff
@@ -571,3 +514,104 @@ def notify_submission_error(instance, message, exec_info=None):
+        self.submit_ip = submit_ip
+        self.path = path
+        self.openrosa_headers = openrosa_headers
+        self.last_sync_token = last_sync_token
+        self.received_on = received_on
+        self.date_header = date_header
+        self.app_id = app_id
+        self.build_id = build_id
+        self.partial_submission = partial_submission
+
+    def process_form(self):
+        self.pre_process_form()
+        result = process_xform_xml(self.domain, self.get_instance(), self.attachments, self.auth_context.to_json())
+        self.post_process_form(result.submitted_form)
+        return result
+
+    def pre_process_form(self):
+        # over write this method to add pre processing to form
+        pass
+
+    def post_process_form(self, xform):
+        self._set_submission_properties(xform)
+        found_old = scrub_meta(xform)
+        legacy_notification_assert(not found_old, 'Form with old metadata submitted', xform.form_id)
+        self._invalidate_caches(xform)
+
+    def _set_submission_properties(self, xform):
+        # attaches shared properties of the request to the document.
+        # used on forms and errors
+        xform.submit_ip = self.submit_ip
+        xform.path = self.path
+
+        xform.openrosa_headers = self.openrosa_headers
+        xform.last_sync_token = self.last_sync_token
+
+        if self.received_on:
+            xform.received_on = self.received_on
+
+        if self.date_header:
+            xform.date_header = self.date_header
+
+        xform.app_id = self.app_id
+        xform.build_id = self.build_id
+        xform.export_tag = ["domain", "xmlns"]
+        xform.partial_submission = self.partial_submission
+
+    def _invalidate_caches(self, xform):
+        for device_id in {None, xform.metadata.deviceID if xform.metadata else None}:
+            self._invalidate_restore_payload_path_cache(xform, device_id)
+            if ASYNC_RESTORE.enabled(self.domain):
+                self._invalidate_async_restore_task_id_cache(xform, device_id)
+
+    def _invalidate_restore_payload_path_cache(self, xform, device_id):
+        """invalidate cached initial restores"""
+        restore_payload_path_cache = RestorePayloadPathCache(
+            domain=self.domain,
+            user_id=xform.user_id,
+            sync_log_id=xform.last_sync_token,
+            device_id=device_id,
+        )
+        restore_payload_path_cache.invalidate()
+
+    def _invalidate_async_restore_task_id_cache(self, xform, device_id):
+        async_restore_task_id_cache = AsyncRestoreTaskIdCache(
+            domain=self.domain,
+            user_id=xform.user_id,
+            sync_log_id=self.last_sync_token,
+            device_id=device_id,
+        )
+
+        task_id = async_restore_task_id_cache.get_value()
+
+        if task_id is not None:
+            revoke_celery_task(task_id)
+            async_restore_task_id_cache.invalidate()
+
+    def get_instance_xml(self):
```

Comment:
this isn't called so should be removed - it can be added to the subclass. Same for `update_instance`

Reply:
these are exposed methods that should be used to update and access the instance. hence the `_instance` property.
They are used in https://github.com/dimagi/commcare-icds/pull/54/files#diff-995b096c82f502a18359635ef379ea24R22-R29

Reply:
if you want to add them to the subclass then you may but they don't belong on this class here and may well get removed in future since they aren't called anywhere in this codebase

Reply:
hmm, that is a valid point that these are not used anywhere in the HQ codebase. The effort here was to have, any custom processor class to follow the same way to get and update instance xml. Currently it's only being used by ICDS, but if later some other environment also adds a custom class, that would need to add such methods again.
Does that make sense?

> may well get removed in future since they aren't called anywhere in this codebase

That might be true for code only used within HQ, but if something is made available for other envs to overrite, should not HQ provide such methods, that are not used with core HQ but are to be used by extensions.

Reply:
lets not over engineer this. I don't foresee any other custom form processors in the future.

Reply:
Updated. https://github.com/dimagi/commcare-hq/pull/28426/commits/4d81fcac48dd5c536d32c42fa91a4acd664cd4ef

Reply:
> That might be true for code only used within HQ, but if something is made available for other envs to overrite, should not HQ provide such methods, that are not used with core HQ but are to be used by extensions.

In this case anything that is specifically for extensions to access it should be clearly documented. I don't want to do that here because we only have a single use case so it seems premature to add it.

Reply:
okay. I was always thinking of providing it in a way, beyond this one use case that we see right now. I don't own how extensions are getting designed so I am going to take a step back on this and go with how you suggest.
Let me know if this looks okay now.